staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -734,16 +769,21 @@ func (alloc *allocator) allocateOne(r deviceIndices, allocateSubRequest bool) (b
 		// Done with request, continue with next one. We have completed the work for
 		// the request or subrequest, so we can no longer be allocating devices for
 		// a subrequest.
-		return alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex + 1}, false)
+		success, err := alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex + 1}, false)
+		// We want to propagate any errAllocationResultMaxSizeExceeded to the caller. If
+		// that error is returned here, it means none of the requests/subrequests after this one
+		// could be allocated while staying within the limit on the number of devices, so there
+		// are no solution in the current request/subrequest that would work.
+		return success, err
 	}
 
 	// Before trying to allocate devices, check if allocating the devices
 	// in the current request will put us over the threshold.
-	numDevicesAfterAlloc := len(alloc.result[r.claimIndex].devices) + requestData.numDevices
+	numDevicesAfterAlloc := len(alloc.result[r.claimIndex].devices) + (requestData.numDevices - r.deviceIndex)
```

Comment:
```suggestion
	numDevicesAfterAlloc := len(alloc.result[r.claimIndex].devices) + requestData.numDevices - r.deviceIndex
```

Reply:
Why substract ` r.deviceIndex`?

Reply:
If `r.deviceIndex` is larger than zero, it means one or more devices have been allocated for the request. Thus, those devices are already accounted for in `len(alloc.result[r.claimIndex].devices)`. So subtracting it is necessary to prevent double counting.